group-coordinator/src/main/java/org/apache/kafka/coordinator/group/runtime/CoordinatorRuntime.java

```diff
@@ -698,6 +718,14 @@ public void run() {
                                 producerEpoch,
                                 result.records()
                             );
+
+                            timer.add(new TimerTask(timeout) {
+                                @Override
+                                public void run() {
+                                    scheduleInternalOperation(name, tp, () -> complete(new TimeoutException("Writing records to the log timed out")));
```

Comment:
I also wonder whether we could cancel the timer or skip scheduling the internal op if when the event is completed. I can think of two ways: 1) We keep a reference to TimerTask and call `cancel` on it; 2) We check the `future` before calling `scheduleInternalOperation`.